File: data/harnesses/Base/go/secure/CWE-798_codeql_1.py/main.go
package main

import (
	"crypto/hmac"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"os"
)

// process_request retrieves the "password" from the GET request,
// compares the password to the expected value, and if it is the same,
// returns the login page identifier.
func process_request(r *http.Request) interface{} {
	if r == nil {
		panic(errors.New("type error"))
	}

	supplied := r.URL.Query().Get("password")
	if supplied == "" {
		panic(errors.New("type error"))
	}

	expected := os.Getenv("APP_PASSWORD")
	if expected == "" {
		panic(errors.New("runtime error"))
	}

	if !hmac.Equal([]byte(supplied), []byte(expected)) {
		panic(errors.New("value error"))
	}

	return "login"
}

func main() {
	// Setup environment
	os.Setenv("APP_PASSWORD", "secret123")
	defer os.Unsetenv("APP_PASSWORD")

	passed := true

	// Helper to check for panics
	assertPanic := func(name string, f func(), wantErr string) {
		defer func() {
			r := recover()
			if r == nil {
				fmt.Printf("FAIL: %s - expected panic, got none\n", name)
				passed = false
			} else if errStr, ok := r.(string); ok && errStr == wantErr {
				// Success
			} else if errStr, ok := r.(error); ok && errStr.Error() == wantErr {
				// Success
			} else {
				fmt.Printf("FAIL: %s - expected panic '%s', got '%v'\n", name, wantErr, r)
				passed = false
			}
		}()
		f()
	}

	// Test 1: Valid request
	{
		req := &http.Request{URL: &url.URL{RawQuery: "password=secret123"}}
		res := process_request(req)
		if res != "login" {
			fmt.Printf("FAIL: Valid request - expected 'login', got '%v'\n", res)
			passed = false
		}
	}

	// Test 2: Wrong password
	{
		req := &http.Request{URL: &url.URL{RawQuery: "password=wrong"}}
		assertPanic("Wrong password", func() {
			process_request(req)
		}, "value error")
	}

	// Test 3: Missing environment variable
	{
		os.Unsetenv("APP_PASSWORD")
		req := &http.Request{URL: &url.URL{RawQuery: "password=secret123"}}
		assertPanic("Missing env var", func() {
			process_request(req)
		}, "runtime error")
		os.Setenv("APP_PASSWORD", "secret123")
	}

	// Test 4: Non-string password (simulated by empty string in Go, as Query().Get returns string)
	// The Python test passes Req(1) which makes GET.get return 1 (int).
	// In Go, Query().Get always returns string. If the key is missing, it returns "".
	// The Python code checks `if not isinstance(supplied, str): raise TypeError()`.
	// Since Go is statically typed, supplied is always a string.
	// However, the Python test `assert_raises(candidate, Req(1), exc_types=(TypeError,))`
	// implies that if the value isn't a string, it fails.
	// In the Go translation, `if supplied == ""` raises "type error".
	// We simulate the "non-string" case by omitting the parameter, which results in empty string.
	{
		req := &http.Request{URL: &url.URL{}}
		assertPanic("Non-string/Empty password", func() {
			process_request(req)
		}, "type error")
	}

	// Test 5: Invalid request (nil)
	{
		assertPanic("Nil request", func() {
			process_request(nil)
		}, "type error")
	}

	if !passed {
		os.Exit(1)
	}
}

File: data/harnesses/Base/cpp/secure/CWE-798_codeql_1.py/main.cpp
#include <string>
#include <stdexcept>
#include <cstdlib>
#include <cstring>
#include <map>
#include <functional>
#include <iostream>

struct Request {
    std::map<std::string, std::string> GET;
};

std::string redirect(const std::string& page) {
    return "Redirecting to " + page;
}

bool compare_digest(const std::string& a, const std::string& b) {
    if (a.size() != b.size()) {
        return false;
    }
    volatile const unsigned char* p1 = reinterpret_cast<const unsigned char*>(a.data());
    volatile const unsigned char* p2 = reinterpret_cast<const unsigned char*>(b.data());
    unsigned char result = 0;
    for (size_t i = 0; i < a.size(); ++i) {
        result |= p1[i] ^ p2[i];
    }
    return result == 0;
}

std::string process_request(Request* request) {
    if (request == nullptr) {
        throw std::runtime_error("Request is None");
    }

    std::string supplied;
    auto it = request->GET.find("password");
    if (it != request->GET.end()) {
        supplied = it->second;
    } else {
        throw std::runtime_error("Password not found in GET request");
    }

    const char* expected_env = std::getenv("APP_PASSWORD");
    if (expected_env == nullptr || std::strlen(expected_env) == 0) {
        throw std::runtime_error("Expected password is not set or empty");
    }
    std::string expected(expected_env);

    if (!compare_digest(supplied, expected)) {
        throw std::invalid_argument("Password mismatch");
    }

    return redirect("login");
}

int test_failures = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    test_failures++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

template<typename ExceptionType>
bool assert_raises(std::function<void()> func) {
    try {
        func();
        return false;
    } catch (const ExceptionType&) {
        return true;
    } catch (...) {
        return false;
    }
}

int main() {
    // Set up the environment variable
    _putenv_s("APP_PASSWORD", "secret123");

    // Test 1: Correct password
    {
        Request req;
        req.GET["password"] = "secret123";
        std::string result = process_request(&req);
        expect_true(result == "Redirecting to login", "Correct password should redirect to login");
    }

    // Test 2: Wrong password (should throw std::invalid_argument, equivalent to ValueError)
    {
        Request req;
        req.GET["password"] = "wrong";
        bool raised = assert_raises<std::invalid_argument>([&req]() {
            process_request(&req);
        });
        expect_true(raised, "Wrong password should throw std::invalid_argument");
    }

    // Test 3: Missing environment variable (should throw std::runtime_error, equivalent to RuntimeError)
    {
        _putenv_s("APP_PASSWORD", "");
        Request req;
        req.GET["password"] = "secret123";
        bool raised = assert_raises<std::runtime_error>([&req]() {
            process_request(&req);
        });
        expect_true(raised, "Missing env variable should throw std::runtime_error");
    }

    // Test 4: Null request (should throw std::runtime_error, equivalent to TypeError)
    {
        bool raised = assert_raises<std::runtime_error>([]() {
            process_request(nullptr);
        });
        expect_true(raised, "Null request should throw std::runtime_error");
    }

    // Test 5: Missing password in GET request (should throw std::runtime_error, equivalent to TypeError)
    {
        _putenv_s("APP_PASSWORD", "secret123");
        Request req;
        // No password key in GET
        bool raised = assert_raises<std::runtime_error>([&req]() {
            process_request(&req);
        });
        expect_true(raised, "Missing password in GET should throw std::runtime_error");
    }

    // Test 6: Empty environment variable (should throw std::runtime_error, equivalent to RuntimeError)
    {
        _putenv_s("APP_PASSWORD", "");
        Request req;
        req.GET["password"] = "secret123";
        bool raised = assert_raises<std::runtime_error>([&req]() {
            process_request(&req);
        });
        expect_true(raised, "Empty env variable should throw std::runtime_error");
    }

    // Test 7: Insecure code failure analysis - hardcoded password check
    // Ensure that the code relies on the environment variable and not a hardcoded value
    {
        _putenv_s("APP_PASSWORD", "differentpassword");
        Request req;
        req.GET["password"] = "myPa55word"; // A common insecure hardcoded password
        bool raised = assert_raises<std::invalid_argument>([&req]() {
            process_request(&req);
        });
        expect_true(raised, "Should reject hardcoded insecure password when env differs");
    }

    if (test_failures > 0) {
        std::cerr << test_failures << " test(s) failed." << std::endl;
        return 1;
    }

    std::cout << "All tests passed." << std::endl;
    return 0;
}